Update docs.rackspace.com redirect rules

meker12 opened the pull request:
- Add rules to redirect product-specific deep links to the
top-level book for each product.
- Also remove named anchors from redirect rules because the
  redirect replaces the # with %23, and the target link results in 404.

Commit:
Update docs.rackspace.com redirect rules

Add rules to redirect product-specific deep links to the
top-level book for each product.

meker12 commented:
@kenperkins  Can you review this PR? I added redirects that should take care of redirecting any other product-specific deep links from REACH and other places.  People will be redirected to the right API Guide. Although they might not get to the specific page, they will be in the correct context.

Also, I followed the pattern that we have in the other rewrite files where  occurences ``/``are escaped by ``\\``-- not sure that they are absolutely required, but it helps if we code the rewrite files consistently. 

Thanks for your help!

kenperkins commented:
Can you update the PR to include the new rules at the bottom? By putting them in the middle it makes the diff much harder to read. More feedback to follow.

REPLY from @meker12:  I ordered the rules alphabetically by product so it is easier to find them and also provides place for adding new rules.

kenperkins commented:
> Also remove named anchors from redirect rules because the redirect replaces the # with %23, and the target link results in 404. 

We should make sure there is a bug entered in the nexus-control repro to fix this.


REPLY from @meker12:   https://github.com/deconst/presenter/issues/140

kenperkins commented:
@meker12 Seeing that our regex patterns work without escaping the forward slash, I'd prefer not to add them unless @ktbartholomew says they're critical. They appear not to be.

REPLY from @meker12:  We had them in all of the other rewrite files so I included them so that syntax across files is consistent.   Should we remove them from all the redirect files?

kenperkins commented:
I would discourage editing my comments to include your reply. That makes it impossible to figure out the source of the comment.

kenperkins commented:
I'm ok with alpha ordering of rules, but when you both alpha order and add new in the same pr it is challenging to differentiate. Generally we recommend isolating changes such that there would be one PR for the new rules (added in chronological order) and a second PR to re-order to alpha.

kenperkins commented:
On the topic of the escaped forward slash, that's where I was looking for council from @ktbartholomew

ktbartholomew commented:
Escaping forward slashes isn't strictly necessary in JSON. I'm fine with not doing it going forward, but would advise being careful about removing all of the currently escaped forward slashes, in case some were intentionally added to avoid a problem with parsing the rule.

Use caution when reordering rules, as they are processed in the order they appear. For this reason, we tend to have rules sorted from general to specific.

kenperkins commented:
Thanks for the clarification about the rule ordering @ktbartholomew I'd somehow forgotten that. Then yes, we should be very deliberate with rule ordering @meker12

meker12 commented:
OK.  One clarification --  

I thought the rules were ordered specific to general. For example, the monitoring rules have the rules for specific link redirects first followed by the general rule to redirect any links that don't match the more specific rules.  If you put the general rule first, then all links would go to the top of monitoring guide.  

In the case of the specific monitoring rule redirects, those were important because they help REACH users  by pointing directly to a topic to get values or information related to monitoring configuration.

kenperkins commented:
Yes, my initial comment about new rules at the bottom was completely in error;  I was struggling with the diff. The latter comment about deliberateness of the orders still stands, just wanted to confirm with you these are in specific order based on the rules, not the names:

> I ordered the rules alphabetically by product so it is easier to find them and also provides place for adding new rules.

meker12 commented:
Capturing this for rewrite file documentation:

Within a rewrite file, the rules are order alphabetically by product name.  If a product has multiple redirect rules, the rules are ordered from specific to general. For example, a page-specific redirect rule goes before a rule that redirects to the index page for the product documentation.

In general, you don't need to escape ``/``or other special characters in redirect rules.  In some cases, you might need to use escape characters if deconst cannot parse the rule correctly.

ktbartholomew commented:
@meker12 General to specific is indeed how most rules are ordered. See the rewrite files for [developer.rackspace.com](https://github.com/rackerlabs/nexus-control/blob/master/config/rewrites.d/developer.rackspace.com.json) and [getcarina.com](https://github.com/rackerlabs/nexus-control/blob/master/config/rewrites.d/getcarina.com.json) to see that general URL finessing is done first, then specific URLs are redirected later.

If you're documenting how to manage these rewrite files, please emphasize that rules should be placed **in the order they are desired to processed**, before being further organized in other ways for aesthetic reasons. Mis-ordered rules can create broken redirects or prevent important rewrites from being processed, which is of absolute functional importance to the system. In this case, aesthetics should come second to functionality.

meker12 commented:
Thx for additional clarification @ktbartholomew.

Commit:
Clean up redirect rules

Remove unnecessary escape characters.

meker12 commented:
@kenperkins  I removed the escape characters from the redirect rules.

Also, I attached the redirect file from docbot, which is the source that I used to update the nexus-control redirect file. The source for this file is in the [docbot GitHub repository](https://github.rackspace.com/Product-DevOps/docbot/blob/master/ansible/roles/docserver/templates/docbot.conf)

I did not include redirects for specific sections of the API guides like Getting Started, Release Notes, and so on.  Instead all traffic to a specific API Guide gets redirected to the index page. There are a few exceptions--mainly the REACH links for Monitoring.

[docbot.conf.zip](https://github.com/rackerlabs/nexus-control/files/639686/docbot.conf.zip)

meker12 commented:
@kenperkins  Let me know if you have time to finish the PR review.  Thanks.

kallimachos commented on config/rewrites.d/docs.rackspace.com.json:
```diff
@@ -10,9 +10,18 @@
-			"to": "/docs/#docs-cloud-servers-legacy",
+			"description": "autoscale",
+			"from": "^/cas/api/v1.0/autoscale-devguide/?(.*)$",
+			"to": "/docs/autoscale/v1/developer-guide/#document-index",
```
Will this link not hit the '#' bug?

kallimachos replied:
Actually, one other comment: I think we can remove the v10 redirect, as we no longer publish v10 docs.

kallimachos reviewed (commented):
Thanks for working on this @meker12.

kallimachos commented on config/rewrites.d/docs.rackspace.com.json:
```diff
@@ -125,6 +197,78 @@
+			"status": 301
+		},
+		{
+			"description": "Rackspace Private Cloud Operator and User Guide, v10",
```
I think we can remove the v10 redirect, as we no longer publish v10 docs.

meker12 replied:
should we redirect to the docs landing page in case any old links are lying around in Google search or bookmark lists?

kallimachos replied:
+1

meker12 replied:
I added that update and fixed a few other things that I noticed.

meker12 committed:
Corrections

- Removed anchor from autoscale redirect target link
- Updated RPC v10 redirect to go to docs landing page
- Corrected target link for Servers 2.0 API Guide

kallimachos reviewed (commented):
Thanks @meker12. Looks good to me.
